src/cls/rbd/cls_rbd.cc

```diff
@@ -229,6 +242,233 @@ static bool is_valid_id(const string &id) {
 
 /**
  * Initialize the header with basic metadata.
+ * Everything is stored as key/value pairs as omaps in the header object.
+ *
+ * If features the OSD does not understand are requested, -ENOSYS is
+ * returned.
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int create_cg(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  bufferlist snap_seqbl;
+  uint64_t snap_seq = 0;
+  ::encode(snap_seq, snap_seqbl);
+  int r = cls_cxx_map_set_val(hctx, CG_SNAP_SEQ, &snap_seqbl);
+  if (r < 0)
+    return r;
+
+  return 0;
+}
+
+int image_add_cg_ref(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "image_add_cg_ref");
+  std::string cg_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(cg_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  bufferlist existing_refbl;
+
+  int r = read_key(hctx, RBD_CG_REF_KEY, &existing_refbl);
+  if (r != -ENOENT) { // No entry means this image is not a member of any cg. So, we can use it.
+    if (r < 0) {
+      return r;
+    }
+
+    std::string old_cg_id;
+    int64_t old_pool_id;
+    try {
+      bufferlist::iterator iter = in->begin();
+      ::decode(old_cg_id, iter);
+      ::decode(old_pool_id, iter);
+    } catch (const buffer::error &err) {
+      return -EINVAL;
+    }
+
+    if ((old_cg_id != cg_id) || (old_pool_id != pool_id)) {
+      return -EEXIST;
+    }
```

Comment:
Why don't you also return immediately when `old_cg_id == cg_id && old_pool_id == pool_id`, instead of updating the map with the same values below?